bindings/rust/bench/benches/constant_time_equals.rs

```diff
@@ -0,0 +1,105 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+//! These benchmarks compare the performance s2n_constant_time_equals with memcmp.
+//! The benchmarks can be run with "cargo bench low-level-comparison"
+//!
+//! The results show that s2n_constant_time_equals is significantly slower than
+//! memcmp, but that for small amounts of data the comparisons remain very cheap,
+//! on the order of hundreds of nano-seconds.
+//!
+//! We also see that for pathological inputs, the overhead of the constant time
+//! equals grows more noticeable, e.g. ~30 μs. This pathological case is roughly
+//! modelled after what might be seen if a client totally fills an extension with
+//! data. For this reason we generally avoid using s2n_constant_time_equals to
+//! compare large amounts of data.
+
+use criterion::{criterion_group, criterion_main, Criterion};
+use rand::RngCore;
+// we don't need any symbols from s2n-tls, but we do need to link against it to
+// make s2n_constant_time_equals available.
+use s2n_tls as _;
+use std::ffi::c_void;
+
+const SMALL_DATA_LENGTH: usize = u8::MAX as usize;
+const MAX_EXTENSION_SIZE: usize = u16::MAX as usize;
+
+// wrapper around the libc memcmp
+fn memcmp(a: &[u8], b: &[u8]) -> i32 {
+    unsafe {
+        libc::memcmp(
+            a.as_ptr() as *const c_void,
+            b.as_ptr() as *const c_void,
+            a.len(),
+        )
+    }
+}
+
+extern "C" {
+    // s2n_constant_time_equals is no exposed publicly, so we manually write the
+    // bindings for it.
+    //
+    // bool s2n_constant_time_equals(const uint8_t *a, const uint8_t *b, const uint32_t len)
+    fn s2n_constant_time_equals(a: *const u8, b: *const u8, len: u32) -> bool;
+}
+
+fn rust_s2n_constant_time_equals(a: &[u8], b: &[u8]) -> bool {
+    unsafe { s2n_constant_time_equals(a.as_ptr(), b.as_ptr(), a.len() as u32) }
+}
+
+fn comparison(criterion: &mut Criterion) {
```

Comment:
What are we trying to do with this PR? Like, it is relevant for #4709, but once that is merged we're not really going to have a use for this comparison. If we're actually trying to benchmark s2n_constant_time_equals(), I would expect to see a test that checks that the comparison of [1, 2, 3, 4, 5] to [1, 0, 0, 0, 0] takes exactly as much time as the comparison of [1, 2, 3, 4, 5] to [1, 2, 3, 4, 5]. But maybe we trust that that case is covered with cbmc?

Reply:
Ya, I see this benchmark function as being strictly interested in the performance of `s2n_constant_time_equals`, and not at all concerned with the correctness of it. We formally assert its correctness with the `ctverif` proofs.